osfoffline/filesystem_manager/sync_local_filesystem_and_db.py

```diff
@@ -11,209 +13,185 @@
             raise FolderNotInFileSystem
 
-        self.osf_path = ProperPath(absolute_osf_dir_path, True)
-        self.observer = observer
-        self.user = user
+        self.osf_path = ProperPath(self.osf_folder, True)
 
-    def emit_new_events(self):
-        for local_child, db_child in self._make_local_db_tuple_list(self.osf_path, self.user):
-            self._emit_single_event_and_recurse(local_child, db_child)
+    def sync_db(self):
+        logger.info('Beginning initial sync')
+        events, nodes = [], [
+            node for node in
+            self.user.top_level_nodes
+            if node.osf_id in self.user.guid_for_top_level_nodes_to_sync
+        ]
+        for node in nodes:
+            logger.info('Resyncing node {}'.format(node))
+            events.extend(self.loop.run_until_complete(self.crawl_node(node)))
+        logger.info('Initial sync finished')
+
+    def match_local_remote(self, local_list, remote_list):
+        ret = {}
+        for local in local_list:
+            ret.setdefault((local.name, local.is_dir), [None, None])[0] = local
+
+        for remote in remote_list:
+            ret.setdefault((remote.name, remote.is_dir), [None, None])[1] = remote
+
+        return ret
+
+    @asyncio.coroutine
+    def crawl_node(self, node, local_path='/', remote_path='/'):
+        logger.debug('Crawling node {} at path {}'.format(node, local_path))
+        events, directories = [], []
+        resources = self.match_local_remote(
+            self.list_local_dir(node, local_path),
+            (yield from self.list_remote_dir(node, remote_path))
+        )
+
+        for (name, is_dir), (local, remote) in resources.items():
+            if is_dir:
+                directories.append((
+                    getattr(local, 'fullpath', None),
+                    getattr(remote, 'osf_id', None)
+                ))
+            event = self.infer_event(local, remote)
+            if event is not None:
+                events.append(event)
+                # DO SOMETHING HERE
+                # yield from self.queue.put(event)
+
+        # yield from self.queue.join()
+
+        for local_dir, remote_dir in directories:
+            yield from self.crawl_node(node, local_dir, remote_dir)
+
+        return events
+
+    @asyncio.coroutine
+    def list_remote_dir(self, node, remote_folder):
+        if remote_folder is None:
+            return []
 
-    def _make_local_db_tuple_list(self, local, db):
-        if local is None and db is None:
-            raise LocalDBBothNone
-        if local and db and self._get_proper_path(local) != self._get_proper_path(db):
-            raise IncorrectLocalDBMatch
-        if local and not isinstance(local, ProperPath):
-            raise InvalidItemType
+        if remote_folder != '/':
+            return (yield from self.osf_query.get_child_files(remote_folder))
 
-        local_children = {}
-        db_children = {}
+        if not hasattr(self, '__nodes'):
```

Comment:
Note to self:  define `self._nodes = None` in the `__init__` method and remove the `hasattr` check.

Otherwise, since we don't reuse the `'LocalDBSync` instance, caching in this way seems fine.